### ISSUE TITLE
memleak produces an invalid BPF program in `alloc_exit`

### ISSUE THREAD

**Issue by goldshtn** [iovisor/bcc#570]
I'm testing memleak with the native stack support and it doesn't pass BPF program verification. I'm getting the following:

```
$ sudo ./memleak.py
Attaching to kmalloc and kfree, Ctrl+C to quit.
bpf: Permission denied
0: (bf) r6 = r1
1: (79) r1 = *(u64 *)(r6 +80)
2: (7b) *(u64 *)(r10 -8) = r1
3: (85) call 14
4: (7b) *(u64 *)(r10 -16) = r0
5: (18) r1 = 0x7b6f4f80
7: (bf) r2 = r10
8: (07) r2 += -16
9: (85) call 1
10: (b7) r1 = 0
11: (7b) *(u64 *)(r10 -24) = r1
12: (7b) *(u64 *)(r10 -32) = r1
13: (7b) *(u64 *)(r10 -40) = r1
14: (15) if r0 == 0x0 goto pc+23
 R0=map_value(ks=8,vs=8) R1=imm0 R6=ctx R10=fp fp-8=pkt_end
15: (79) r1 = *(u64 *)(r0 +0)
16: (7b) *(u64 *)(r10 -40) = r1
17: (18) r1 = 0x7b6f4f80
19: (bf) r2 = r10
20: (07) r2 += -16
21: (85) call 3
22: (85) call 5
23: (7b) *(u64 *)(r10 -32) = r0
24: (18) r2 = 0x7ae00000
26: (bf) r1 = r6
27: (b7) r3 = 1024
28: (85) call 27
29: (63) *(u32 *)(r10 -24) = r0
30: (18) r1 = 0x7b6f4200
32: (bf) r2 = r10
33: (07) r2 += -8
34: (bf) r3 = r10
35: (07) r3 += -40
36: (b7) r4 = 0
37: (85) call 2
invalid indirect read from stack off -8+0 size 8

Traceback (most recent call last):
  File "./memleak.py", line 263, in <module>
    bpf_program.attach_kretprobe(event="__kmalloc", fn_name="alloc_exit")
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 417, in attach_kretprobe
    fn = self.load_func(fn_name, BPF.KPROBE)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 219, in load_func
    raise Exception("Failed to load BPF program %s" % func_name)
Exception: Failed to load BPF program alloc_exit
```

This is with the latest bcc master, on FC22:

```
$ uname -r
4.7.0-0.rc3.git0.1.vanilla.knurd.1.fc22.x86_64
```

Interestingly, this is also happening with the old/memleak.py, so it might be unrelated to the memleak changes but rather to changes in the clang rewriter. Or possibly the kernel? I will try to downgrade to 4.6 and test there.

**Comment by goldshtn** [comment]
Just to add, the error seems to be stemming from the use of `address` in updating the map in `alloc_exit`. Here is the `alloc_exit` probe:

```c
int alloc_exit(struct pt_regs *ctx)
{
        u64 address = PT_REGS_RC(ctx);
        u64 pid = bpf_get_current_pid_tgid();
        u64* size64 = sizes.lookup(&pid);
        struct alloc_info_t info = {0};

        if (size64 == 0)
                return 0; // missed alloc entry

        info.size = *size64;
        sizes.delete(&pid);

        info.timestamp_ns = bpf_ktime_get_ns();
        info.stack_id = stack_traces.get_stackid(ctx, STACK_FLAGS);
        allocs.update(&address, &info);

        if (SHOULD_PRINT) {
                bpf_trace_printk("alloc exited, size = %lu, result = %lx\\n",
                                 info.size, address);
        }
        return 0;
}
```

Removing the map update (`allocs.update(...)`) makes the program validate successfully. Also, changing `u64 address = PT_REGS_RC(ctx)` to `u64 address = 12` makes the program validate successfully. @drzaeus77 it would be great if you could take a look...

**Comment by 4ast** [comment]
the existing memleak.py works fine 4.7.0-rc1 and clang-3.9.
Do you have any changes to memleak.py? what is the llvm/clang version?
The asm looks fine and I'm not sure why verifier complains, but I need to reproduce first to debug it further.

**Comment by 4ast** [comment]
if it's 3.7.0 then it's definitely buggy. Please use 3.7.1 or later.

**Comment by goldshtn** [comment]
Oh sorry for not answering the first part of your question. I don't have any changes to memleak, it's the master version.

**Comment by goldshtn** [comment]
The build instructions say use 3.7.0 :)
I'll try 3.7.1 and update. Although -- if the assembly looks ok, how would updating clang help?

**Comment by 4ast** [comment]
it's hard to follow the asm. It 'looks' ok ;)

**Comment by goldshtn** [comment]
Yeah, so I tried 3.7.1 and am getting the same error. Should I try to downgrade to 4.6? Is there anything I can do on my end that would help diagnose the verifier error?

**Comment by goldshtn** [comment]
I commented out most of the code, and even this version causes a verifier error:

```c
int alloc_exit(struct pt_regs *ctx)
{
        u64 address = PT_REGS_RC(ctx);
        struct alloc_info_t info = {};
        allocs.update(&address, &info);
        return 0;
}
```

Result:

```
$ sudo ./memleak.py
Attaching to kmalloc and kfree, Ctrl+C to quit.
bpf: Permission denied
0: (79) r1 = *(u64 *)(r1 +80)
1: (7b) *(u64 *)(r10 -8) = r1
2: (b7) r1 = 0
3: (7b) *(u64 *)(r10 -16) = r1
4: (7b) *(u64 *)(r10 -24) = r1
5: (7b) *(u64 *)(r10 -32) = r1
6: (18) r1 = 0x7b6f4380
8: (bf) r2 = r10
9: (07) r2 += -8
10: (bf) r3 = r10
11: (07) r3 += -32
12: (b7) r4 = 0
13: (85) call 2
invalid indirect read from stack off -8+0 size 8

Traceback (most recent call last):
  File "./memleak.py", line 264, in <module>
    bpf_program.attach_kretprobe(event="__kmalloc", fn_name="alloc_exit")
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 417, in attach_kretprobe
    fn = self.load_func(fn_name, BPF.KPROBE)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 219, in load_func
    raise Exception("Failed to load BPF program %s" % func_name)
Exception: Failed to load BPF program alloc_exit
```

**Comment by goldshtn** [comment]
Yeah, so I can repro this stand-alone. Here goes:

```python
from bcc import BPF

program = BPF(text="""
#include <linux/ptrace.h>

struct alloc_info_t
{
    u64 size;
};

BPF_HASH(allocs, u64, struct alloc_info_t);

int alloc_exit(struct pt_regs *ctx)
{
    u64 address = PT_REGS_RC(ctx);
    struct alloc_info_t info = {};
    allocs.update(&address, &info);
    return 0;
}
""")

program.attach_kretprobe(event="__kmalloc", fn_name="alloc_exit")
```

Causes the same verifier error:

```
$ sudo python ~/test.py
bpf: Permission denied
0: (79) r1 = *(u64 *)(r1 +80)
1: (7b) *(u64 *)(r10 -8) = r1
2: (b7) r1 = 0
3: (7b) *(u64 *)(r10 -16) = r1
4: (18) r1 = 0x7a5ad700
6: (bf) r2 = r10
7: (07) r2 += -8
8: (bf) r3 = r10
9: (07) r3 += -16
10: (b7) r4 = 0
11: (85) call 2
invalid indirect read from stack off -8+0 size 8

Traceback (most recent call last):
  File "/home/vagrant/test.py", line 22, in <module>
    program.attach_kretprobe(event="__kmalloc", fn_name="alloc_exit")
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 417, in attach_kretprobe
    fn = self.load_func(fn_name, BPF.KPROBE)
  File "/usr/lib/python2.7/site-packages/bcc/__init__.py", line 219, in load_func
    raise Exception("Failed to load BPF program %s" % func_name)
Exception: Failed to load BPF program alloc_exit
```

**Comment by goldshtn** [comment]
More data: if I use `PT_REGS_PARM1(ctx)` there's no verifier error. 😿

**Comment by 4ast** [comment]
thanks for narrowing it down.
yes. it's a kernel bug.
will fix asap.

**Comment by goldshtn** [comment]
OK cool! Was it introduced between rc1 and rc3?

**Comment by 4ast** [comment]
it's broken in all 4.7. It was accidentally fixed in the stuff we're planning to do in 4.8.
i'll send the fix soon.
